packages/angular/src/utils/mfe/with-module-federation.ts

```diff
@@ -0,0 +1,222 @@
+  readCachedProjectGraph,
+} from '@nrwl/workspace/src/core/project-graph';
+import { readWorkspaceJson } from '@nrwl/workspace';
+import { joinPathFragments, ProjectGraph } from '@nrwl/devkit';
+import ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');
+import { Workspaces } from 'nx/src/shared/workspace';
+import { appRootPath } from 'nx/src/utils/app-root';
+import { getRootTsConfigPath } from '@nrwl/workspace/src/utilities/typescript';
+import { readTsConfig } from '@nrwl/workspace';
+import { ParsedCommandLine } from 'typescript';
+
+export type MFERemotes = string[] | [remoteName: string, remoteUrl: string][];
+
+export interface MFEConfig {
+  name: string;
+  remotes?: MFERemotes;
+  exposes?: Record<string, string>;
+  shared?: (
+    libraryName: string,
+    library: SharedLibraryConfig
+  ) => SharedLibraryConfig | false;
+}
+
+function recursivelyResolveWorkspaceDependents(
+  projectGraph: ProjectGraph<any>,
+  target: string
+) {
+  let dependencies = [target];
+
+  const workspaceDependencies = (
+    projectGraph.dependencies[target] ?? []
+  ).filter((dep) => !dep.target.startsWith('npm:'));
+  if (workspaceDependencies.length > 0) {
+    for (const dep of workspaceDependencies) {
+      dependencies = [
+        ...dependencies,
+        ...recursivelyResolveWorkspaceDependents(projectGraph, dep.target),
+      ];
+    }
+  }
+  return dependencies;
+}
+
+function mapWorkspaceLibrariesToTsConfigImport(workspaceLibraries: string[]) {
+  const { projects } = new Workspaces(appRootPath).readWorkspaceConfiguration();
+  const tsConfigPath = process.env.NX_TSCONFIG_PATH ?? getRootTsConfigPath();
+  const tsConfig: ParsedCommandLine = readTsConfig(tsConfigPath);
+
+  const tsconfigPathAliases: Record<string, string[]> = tsConfig.options?.paths;
+
+  if (!tsconfigPathAliases) {
+    return workspaceLibraries;
+  }
+
+  const mappedLibraries = [];
+  for (const lib of workspaceLibraries) {
+    const sourceRoot = projects[lib].sourceRoot;
+    let found = false;
+
+    for (const [key, value] of Object.entries(tsconfigPathAliases)) {
+      if (value.find((p) => p.startsWith(sourceRoot))) {
+        mappedLibraries.push(key);
+        found = true;
+        break;
+      }
+    }
+
+    if (!found) {
+      mappedLibraries.push(lib);
+    }
+  }
+
+  return mappedLibraries;
+}
+
+async function getDependentPackagesForProject(name: string) {
+  let projectGraph: ProjectGraph<any>;
+  try {
+    projectGraph = readCachedProjectGraph();
+  } catch (e) {
+    projectGraph = await createProjectGraphAsync();
+  }
+
+  const deps = projectGraph.dependencies[name].reduce(
+    (dependencies, dependency) => {
+      const workspaceLibraries = dependencies.workspaceLibraries;
+      const npmPackages = dependencies.npmPackages;
+
+      if (dependency.target.startsWith('npm')) {
+        npmPackages.push(dependency.target.replace('npm:', ''));
+      } else {
+        workspaceLibraries.push(dependency.target);
```

Comment:
Both, the npm packages and the workspace libraries might end up duplicated, consider using a `Set` to avoid duplication. It's not a functional issue in the current usage because later these arrays are turned into objects where the duplication will be overwritten, but's still more processing to do that can be avoided.